experimental/dds/tree2/src/shared-tree-core/branch.ts

```diff
@@ -218,7 +218,12 @@ export class SharedTreeBranch<TEditor extends ChangeFamilyEditor, TChange> exten
 			// state of the branch at the start of the transaction.
 			this.repairDataStoreProvider.freeze();
 		}
-		this.transactions.push(this.head.revision, repairStore);
+		const forks = new Set<SharedTreeBranch<TEditor, TChange>>();
+		onForkTransitive(this, (fork) => {
```

Comment:
One other thing to consider is whether or not you want to deregister all of the `fork` events when the transaction ends. Normally it would be good practice to do that, but in this case it might not matter so much, since all the forks are being disposed, and likely dropped from memory. In which case, it doesn't matter if they keep around an event registration because that registration will also be dropped. However, in the case where a user creates a large number of branches during a transaction and keeps references to them even after they are disposed, it would be beneficial to deregister the events. On the one hand, that seems like a fairly unlikely scenario, but on the other hand, deregistering the events is not very hard, so I'll let you decide if you want to do it or not.

Reply:
I think this adds 2 events subscription per fork, plus the ones added in `onForkTransitive` are pushed to an array, so I don't think the memory overhead is not that much of a deal.

Reply:
But anylizing it a bit deep its better to address this because the memory compexity from the use of `onForkTransitive` is O(h^2).

Reply:
My thought was that for a given tree of forked branches, each branch will retain:

* An event registration (stored in its event emitter). The aggregate cost is O(n) for n branches.
* A reference to the deregistration list (`offs`) belonging to its parent (captured via closure in fork event handler). This aggregate cost is also O(n).
* A reference to its own `offs`, which grows according to the number of forks it creates. Each branch can have multiple forks, so in theory the cost is O(n^2), or O(n*f) or something. However, we know that the total number of forks is bounded by n, so it's still just O(n).

I might be missing another cost - how are you arriving at O(n^2)? Or did you mean `h` as in "height"?

Reply:
I mean "height", but at the end is quadratic, so it should be released.

Reply:
My calculation is the worst case scenario that every fork created is a child of the previous one and each of them have at least a transaction.

Reply:
I see - it is quadratic over the number of calls to `onForkTransitive` (`onForkTransitive` itself does not take quadratic memory). How would you propose optimizing that scenario?

Reply:
I think a way to optimize this is not using `onForkTransitive` and just be able to execute dispose recursively for a brach and ist children, this way every branch/fork only stores a reference of its direct children.